aten/src/ATen/native/mps/operations/Copy.mm

```diff
@@ -0,0 +1,237 @@
+      id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
+      id<MTLBlitCommandEncoder> blitEncoder =
+          [commandBuffer blitCommandEncoder];
+
+      [blitEncoder copyFromBuffer:sourceBuffer
+                     sourceOffset:0
+                         toBuffer:destBuffer
+                destinationOffset:0
+                             size:size];
+      [blitEncoder endEncoding];
+      stream->commitAndWait();
+    }
+  });
+}
+
+
+static at::Tensor& copy_kernel_mps(at::Tensor& dst, const at::Tensor& src,
+                            bool non_blocking) {
+  MPSStream* stream = getCurrentMPSStream();
+  uint64_t size = src.nbytes();
+
+  auto src_byte_offset = src.storage_offset() * src.itemsize();
+  id<MTLBuffer> sourceBuffer = __builtin_bit_cast(id<MTLBuffer>, src.storage().data());
+
+  auto dst_byte_offset = dst.storage_offset() * dst.itemsize();
+  id<MTLBuffer> destBuffer = __builtin_bit_cast(id<MTLBuffer>, dst.storage().data());
+
+  dispatch_sync(stream->queue(), ^() {
+    @autoreleasepool {
+      id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
+      id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer blitCommandEncoder];
+
+      [blitEncoder copyFromBuffer:sourceBuffer
+                     sourceOffset:src_byte_offset
+                         toBuffer:destBuffer
+                destinationOffset:dst_byte_offset
+                             size:size];
+      [blitEncoder endEncoding];
+      if (non_blocking) {
+        stream->commit(true);
+      } else {
+        stream->commitAndWait();
+      }
+    }
+  });
+  return dst;
+}
+
+at::Tensor& mps_copy_(at::Tensor& dst, const at::Tensor& src, bool non_blocking)
+{
+  TORCH_CHECK(dst.defined(), "dst is undefined");
+  TORCH_CHECK(src.defined(), "src is undefined");
+
+  if (src.numel() == 0 || dst.is_same(src)) {
+    return dst;
+  }
+  if (dst.numel() == 0) {
+    dst.resize_as_(src);
+  }
+
+  if (src.device().type() == at::kMPS && dst.device().type() == at::kCPU) {
+    return copy_from_mps_(dst, src, non_blocking);
+  }
+  if (src.device().type() == at::kCPU && dst.device().type() == at::kMPS) {
+    return copy_to_mps_(dst, src, non_blocking);
+  }
+
+  if (src.device().type() == at::kMPS && dst.device().type() == at::kMPS) {
+    return copy_kernel_mps(dst, src, non_blocking);
+  }
+  TORCH_INTERNAL_ASSERT(
+      src.device().type() == DeviceType::MPS,
+      "mps_copy_ is implemented only for *->MPS; MPS->*");
+  return dst;
+}
+
+at::Tensor _copy_from_and_resize(const at::Tensor& self, const at::Tensor& dst)
+{
+  return mps_copy_(const_cast<Tensor&>(dst), self, false);
+}
+
+at::Tensor _copy_from(const at::Tensor& self, const at::Tensor& dst, bool non_blocking)
+{
+  return mps_copy_(const_cast<Tensor&>(dst), self, non_blocking);
+}
+
+} // namespace mps
+
+TORCH_LIBRARY_IMPL(aten, MPS, m) {
+   m.impl("_copy_from", TORCH_FN(mps::_copy_from));
+   m.impl("_copy_from_and_resize", TORCH_FN(mps::_copy_from_and_resize));
```

Comment:
Any reason not to define these two kernels in the `at::native::` namespace? Then you can add these two entries directly in native_functions.yaml (by adding an `MPS` for these two functions).

Then you can avoid the `TORCH_LIBRARY_IMPL` boilerplate here, and make it clearer to people reading `native_functions.yaml` which operators have MPS support.

Reply:
Good catch, this was overlooked from our side. We had started initially with `LIBRARY_IMPL` but quickly moved to `native_functions.yaml` paradigm. This should indeed be moved to that place.

Reply:
Hi @bdhirsh 
There is no dispatch backend specific impl for these _copy and _copy_and_resize

```
- func: _copy_from(Tensor self, Tensor dst, bool non_blocking=False) -> Tensor
  dispatch: {}

# We need this to be able to properly copy from a CPU to an XLA tensor with different sizes.
# See https://github.com/pytorch/xla/issues/2881
- func: _copy_from_and_resize(Tensor self, Tensor dst) -> Tensor
  dispatch: {}
```

Where are these functions declared and implemented?

Reply:
Done.